pkg/controller/periodic/periodic.go

```diff
@@ -24,101 +23,97 @@ type Configurator interface {
+	for key := range c.statuses {
 		keys = append(keys, key)
 	}
 	sort.Strings(keys)
 	sources := make([]controllerstatus.Interface, 0, len(keys))
 	for _, key := range keys {
-		sources = append(sources, c.status[key])
+		sources = append(sources, c.statuses[key])
 	}
 	return sources
 }
 
-// sync gathers data from the cluster periodically
-func (c *Controller) sync(name string) error {
+func (c *Controller) Run(stopCh <-chan struct{}, initialDelay time.Duration) {
+	defer utilruntime.HandleCrash()
+	defer klog.Info("Shutting down")
+	c.initialDelay = initialDelay
+
+	go wait.Until(func() { c.periodicTrigger(stopCh) }, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (c *Controller) Gather() {
+	for name := range c.gatherers {
+		start := time.Now()
+		err := c.runGatherer(name)
+		if err == nil {
+			klog.V(4).Infof("Periodic gather %s completed in %s", name, time.Since(start).Truncate(time.Millisecond))
+			c.statuses[name].UpdateStatus(controllerstatus.Summary{Healthy: true})
+			return
+		}
+		utilruntime.HandleError(fmt.Errorf("%v failed after %s with: %v", name, time.Since(start).Truncate(time.Millisecond), err))
+		c.statuses[name].UpdateStatus(controllerstatus.Summary{Reason: "PeriodicGatherFailed", Message: fmt.Sprintf("Source %s could not be retrieved: %v", name, err)})
+	}
+}
+
+func (c *Controller) runGatherer(name string) error {
 	gatherer, ok := c.gatherers[name]
 	if !ok {
 		klog.V(2).Infof("No such gatherer %s", name)
 		return nil
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), c.config.Config().Interval/2)
+	ctx, cancel := context.WithTimeout(context.Background(), c.configurator.Config().Interval/2)
 	defer cancel()
 	defer func() {
 		if err := c.recorder.Flush(ctx); err != nil {
 			klog.Errorf("Unable to flush recorder: %v", err)
 		}
 	}()
 	klog.V(4).Infof("Running %s", name)
-	return gatherer.Gather(ctx, c.config.Config().Gather, c.recorder)
-}
-
-// Run starts gathering with initialDelay
-func (c *Controller) Run(workers int, stopCh <-chan struct{}, initialDelay time.Duration) {
-	defer utilruntime.HandleCrash()
-	defer c.queue.ShutDown()
-	defer klog.Info("Shutting down")
-
-	// start watching for version changes
-	go wait.Until(func() { c.periodicTrigger(stopCh) }, time.Second, stopCh)
-
-	for i := 0; i < workers; i++ {
-		go wait.Until(func() {
-			if initialDelay > 0 {
-				select {
-				case <-stopCh:
-					return
-				case <-time.After(initialDelay):
-					c.runWorker()
-				}
-			}
-			c.runWorker()
-		}, time.Second, stopCh)
-	}
-
-	// seed the queue
-	c.Gather()
-
-	<-stopCh
-}
-
-func (c *Controller) Gather() {
-	for name := range c.gatherers {
-		c.queue.Add(name)
-	}
+	return gatherer.Gather(ctx, c.configurator.Config().Gather, c.recorder)
 }
 
 func (c *Controller) periodicTrigger(stopCh <-chan struct{}) {
-	configCh, closeFn := c.config.ConfigChanged()
+	configCh, closeFn := c.configurator.ConfigChanged()
 	defer closeFn()
 
-	interval := c.config.Config().Interval
+	if c.initialDelay > 0 {
```

Comment:
What happens if the `initialDelay = 0`? Will the gathering be executed immediately? It doesn't seem so....but maybe I am missing something

Reply:
Yep. It looks weird. AFIK, the Gather will not be executed if the `initialDelay` is less or eq to zero. I think that this behavior is not desired. Can we put some default value for it?

Reply:
`InitialDelay` has the default value 0 - https://github.com/openshift/insights-operator/blob/master/pkg/controller/operator.go#L63

Reply:
if `initialDelay == 0` then the gathering will start as normal at https://github.com/openshift/insights-operator/pull/302/files#diff-dec12f0cf28a91a837b46e755b46c081c9014121cdb770b2cd8ac916af29c8deR133

Reply:
or do we want to start a gather immediately if no `initialDelay` is set

Reply:
@0sewa0 yes I think so

Reply:
updated it
it should run a gather immediately if `initialDelay` is 0 or less

Reply:
thanks